# Autos Project

This project works with a dataset of used cars from eBay Kleinanzeigen (https://data.world/data-society/used-cars-data). We're going to clean up the data set and then try to extract some useful info from it.

The data dictionary provided with data is as follows:

dateCrawled - When this ad was first crawled. All field-values are taken from this date.

name - Name of the car.

seller - Whether the seller is private or a dealer.

offerType - The type of listing

price - The price on the ad to sell the car.

abtest - Whether the listing is included in an A/B test.

vehicleType - The vehicle Type.

yearOfRegistration - The year in which which year the car was first registered.

gearbox - The transmission type.

powerPS - The power of the car in PS.

model - The car model name.

kilometer - How many kilometers the car has driven.

monthOfRegistration - The month in which which year the car was first registered.

fuelType - What type of fuel the car uses.

brand - The brand of the car.

notRepairedDamage - If the car has a damage which is not yet repaired.

dateCreated - The date on which the eBay listing was created.

nrOfPictures - The number of pictures in the ad.

postalCode - The postal code for the location of the vehicle.

lastSeenOnline - When the crawler saw this ad last online.

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos=pd.read_csv('Datasets/autos.csv', encoding='Latin-1' )

In [3]:
autos

dateCrawled                                          name  \
0       2016-03-24 11:52:17                                    Golf_3_1.6   
1       2016-03-24 10:58:45                          A5_Sportback_2.7_Tdi   
2       2016-03-14 12:52:21                Jeep_Grand_Cherokee_"Overland"   
3       2016-03-17 16:54:04                            GOLF_4_1_4__3TÜRER   
4       2016-03-31 17:25:20                Skoda_Fabia_1.4_TDI_PD_Classic   
...                     ...                                           ...   
371523  2016-03-14 17:48:27                    Suche_t4___vito_ab_6_sitze   
371524  2016-03-05 19:56:21         Smart_smart_leistungssteigerung_100ps   
371525  2016-03-19 18:57:12            Volkswagen_Multivan_T4_TDI_7DC_UY2   
371526  2016-03-20 19:41:08                        VW_Golf_Kombi_1_9l_TDI   
371527  2016-03-07 19:39:19  BMW_M135i_vollausgestattet_NP_52.720____Euro   

        seller offerType  price   abtest vehicleType  yearOfRegistration  \
0       privat   Angebot    480     test         NaN                1993   
1       privat   Angebot  18300     test       coupe                2011   
2       privat   Angebot   9800     test         suv                2004   
3       privat   Angebot   1500     test  kleinwagen                2001   
4       privat   Angebot   3600     test  kleinwagen                2008   
...        ...       ...    ...      ...         ...                 ...   
371523  privat   Angebot   2200     test         NaN                2005   
371524  privat   Angebot   1199     test      cabrio                2000   
371525  privat   Angebot   9200     test         bus                1996   
371526  privat   Angebot   3400     test       kombi                2002   
371527  privat   Angebot  28990  control   limousine                2013   

          gearbox  powerPS        model  kilometer  monthOfRegistration  \
0         manuell        0         golf     150000                    0   
1         manuell      190          NaN     125000                    5   
2       automatik      163        grand     125000                    8   
3         manuell       75         golf     150000                    6   
4         manuell       69        fabia      90000                    7   
...           ...      ...          ...        ...                  ...   
371523        NaN        0          NaN      20000                    1   
371524  automatik      101       fortwo     125000                    3   
371525    manuell      102  transporter     150000                    3   
371526    manuell      100         golf     150000                    6   
371527    manuell      320      m_reihe      50000                    8   

       fuelType           brand notRepairedDamage          dateCreated  \
0        benzin      volkswagen               NaN  2016-03-24 00:00:00   
1        diesel            audi                ja  2016-03-24 00:00:00   
2        diesel            jeep               NaN  2016-03-14 00:00:00   
3        benzin      volkswagen              nein  2016-03-17 00:00:00   
4        diesel           skoda              nein  2016-03-31 00:00:00   
...         ...             ...               ...                  ...   
371523      NaN  sonstige_autos               NaN  2016-03-14 00:00:00   
371524   benzin           smart              nein  2016-03-05 00:00:00   
371525   diesel      volkswagen              nein  2016-03-19 00:00:00   
371526   diesel      volkswagen               NaN  2016-03-20 00:00:00   
371527   benzin             bmw              nein  2016-03-07 00:00:00   

        nrOfPictures  postalCode             lastSeen  
0                  0       70435  2016-04-07 03:16:57  
1                  0       66954  2016-04-07 01:46:50  
2                  0       90480  2016-04-05 12:47:46  
3                  0       91074  2016-03-17 17:40:17  
4                  0       60437  2016-04-06 10:17:21  
...              ...         ...                  ..

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

In [5]:
autos.info(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

Our dataset contains 20 columns and a whopping 370 thousand rows, most of which are stored as strings. There are some columns that contain dates stored as strings.

We'll start by cleaning the column names to make the data easier to work with (changing it from CamelCase to snakecase).

In [6]:
autos.columns


Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
#correcting some column names

autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gear_box', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']

In [8]:
autos.head()

date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price ab_test vehicle_type  registration_year   gear_box  power_PS  model  \
0    480    test          NaN               1993    manuell         0   golf   
1  18300    test        coupe               2011    manuell       190    NaN   
2   9800    test          suv               2004  automatik       163  grand   
3   1500    test   kleinwagen               2001    manuell        75   golf   
4   3600    test   kleinwagen               2008    manuell        69  fabia   

   odometer  registration_month fuel_type       brand unrepaired_damage  \
0    150000                   0    benzin  volkswagen               NaN   
1    125000                   5    diesel        audi                ja   
2    125000                   8    diesel        jeep               NaN   
3    150000                   6    benzin  volkswagen              nein   
4     90000                   7    diesel       skoda              nein   

            ad_created  num_photos  postal_code            last_seen  
0  2016-03-24 00:00:00           0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00           0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00           0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00           0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00           0        60437  2016-04-06 10:17:21

# Initial Data Exploration and Cleaning

Exploring the dataset to discover obvious areas we can clean


For this dataset we will look for: - Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.

In [9]:
autos.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

       ab_test vehicle_type  registration_year gear_box       power_PS  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model       odometer  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  num_photos   postal_code  \
count             299468               371528    371528.0  371528.00000   
unique                 2                  114         NaN           NaN   
top                 nein  2016-04-03 00:00:00         NaN           NaN   
freq              263182                14450         NaN           NaN   
mean                 NaN                  NaN         0.0   50820.66764   
std                  NaN                  NaN         0.0   25799.08247   
min                  NaN                  NaN         0.0    1067.00000   
25%                  NaN                  NaN         0.0   30459.00000   
50%                  NaN                  NaN         0.0   49610.00000   
75%                  NaN                  NaN         0.0   71546.00000   
max                  NaN                  NaN         0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

Our initial observations:

1. There are a number of text columns where all (or nearly all) of the values are the same:
i. seller
ii. offer_type

2. The num_photos column looks odd, we'll need to investigate this further.

In [10]:
autos['num_photos'].value_counts()

0    371528
Name: num_photos, dtype: int64

It looks like the num_photos column has 0 for every column. We'll drop this column, plus the other two we noted as mostly one value.

In [11]:
autos = autos.drop(["num_photos", "seller", "offer_type"], axis=1)


There are two columns, price and auto, which are numeric values with extra characters being stored as text. We'll clean and convert these.

In [12]:
autos["price"].value_counts()

0         10778
500        5670
1500       5394
1000       4649
1200       4594
          ...  
23456         1
171000        1
21830         1
13485         1
8188          1
Name: price, Length: 5597, dtype: int64

We'll need to clean the dataset to reduce our workload and make the dataset easier to work with

In [13]:
autos = autos.dropna()

autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 260956 entries, 3 to 371527
Data columns (total 17 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   date_crawled        260956 non-null  object
 1   name                260956 non-null  object
 2   price               260956 non-null  int64 
 3   ab_test             260956 non-null  object
 4   vehicle_type        260956 non-null  object
 5   registration_year   260956 non-null  int64 
 6   gear_box            260956 non-null  object
 7   power_PS            260956 non-null  int64 
 8   model               260956 non-null  object
 9   odometer            260956 non-null  int64 
 10  registration_month  260956 non-null  int64 
 11  fuel_type           260956 non-null  object
 12  brand               260956 non-null  object
 13  unrepaired_damage   260956 non-null  object
 14  ad_created          260956 non-null  object
 15  postal_code         260956 non-null  int64 
 16  la

There is uniformity in the data now and we can continue our cleaning

In [14]:
autos["price"].describe()

count    2.609560e+05
mean     8.208628e+03
std      3.433091e+05
min      0.000000e+00
25%      1.500000e+03
50%      3.850000e+03
75%      8.600000e+03
max      1.000000e+08
Name: price, dtype: float64

In [15]:
#changing the price column from float64 to int to make it easier to work with
autos["price"] = (autos["price"].astype(int)
                          )
autos["price"].head()

3    1500
4    3600
5     650
6    2200
7       0
Name: price, dtype: int32

In [16]:
#changing the odometer column from float64 to int and renaming it
autos["odometer"] = (autos["odometer"].astype(int)
                             )
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)
autos["odometer_km"].head()

3    150000
4     90000
5    150000
6    150000
7     40000
Name: odometer_km, dtype: int32

Let's continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the odometer_km and price columns. 

In [17]:

autos["odometer_km"].value_counts()

150000    161236
125000     28761
100000     11995
90000       9934
80000       8907
70000       7974
60000       7229
50000       6368
40000       5364
30000       4945
20000       4358
5000        2406
10000       1479
Name: odometer_km, dtype: int64


We can see that the values in this field are rounded, which might indicate that sellers had to choose from pre-set options for this field. Additionally, there are more high mileage than low mileage vehicles.

In [18]:

print(autos["price"].unique().shape)
print(autos["price"].describe())
print(autos["price"].value_counts().sort_index(ascending=False
                                        ).head(5))

print(autos["price"].value_counts().sort_index(ascending=True
                                        ).head(5))

(5137,)
count    2.609560e+05
mean     8.208628e+03
std      3.433091e+05
min      0.000000e+00
25%      1.500000e+03
50%      3.850000e+03
75%      8.600000e+03
max      1.000000e+08
Name: price, dtype: float64
99999999    2
99000000    1
14000500    1
12345678    2
11111111    1
Name: price, dtype: int64
0    3389
1     340
2       4
3       2
5       5
Name: price, dtype: int64


There are 3,389 cars listed with $0 price - given that this is less than 0.1% of the of the cars, we will be removing these rows.

Also, there are some major outliers in the price coulumn with prices ranging from ridiculously low amounts (i.e $1, $2 & $3) to extremely high amounts (i.e $99999999). 

To work with this dataset, we would need to set a range for the price

In [19]:
#Streamlining the entire dataset to only include cars wiith prices between $10,000-$1,000,000

autos = autos[autos['price'].between(10000, 1000000)]

autos["price"].describe()

count      53323.000000
mean       19544.058811
std        19976.370507
min        10000.000000
25%        12500.000000
50%        15700.000000
75%        21625.000000
max      1000000.000000
Name: price, dtype: float64

# Exploring the date columns

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:


- `date_crawled`:  added by the crawler
- `last_seen`:  added by the crawler
- `ad_created`:  from the website
- `registration_month`:  from the website
- `registration_year`:  from the website

Three of the columns above are stored string values and two are stored as numeric values

In [20]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
14  2016-03-21 12:57:01  2016-03-21 00:00:00  2016-04-06 07:45:42
20  2016-04-01 19:56:48  2016-04-01 00:00:00  2016-04-05 16:45:49
33  2016-03-25 21:48:47  2016-03-25 00:00:00  2016-04-07 04:45:21
38  2016-04-01 17:45:07  2016-04-01 00:00:00  2016-04-05 12:47:30
49  2016-04-04 14:06:22  2016-04-04 00:00:00  2016-04-05 12:09:29

In [21]:
#the first 10 characters represent the day  (e.g. 2016-03-12)

print(autos['date_crawled'].str[:10].head())

14    2016-03-21
20    2016-04-01
33    2016-03-25
38    2016-04-01
49    2016-04-04
Name: date_crawled, dtype: object


In [22]:

(autos["date_crawled"]
        .str[:10]
        .value_counts(dropna=False)
        .sort_index()
        )

2016-03-05    1366
2016-03-06     789
2016-03-07    1892
2016-03-08    1658
2016-03-09    1740
2016-03-10    1609
2016-03-11    1735
2016-03-12    1940
2016-03-13     922
2016-03-14    1916
2016-03-15    1729
2016-03-16    1506
2016-03-17    1559
2016-03-18     661
2016-03-19    1837
2016-03-20    1956
2016-03-21    1819
2016-03-22    1596
2016-03-23    1692
2016-03-24    1662
2016-03-25    1780
2016-03-26    1750
2016-03-27    1658
2016-03-28    1888
2016-03-29    1816
2016-03-30    1768
2016-03-31    1725
2016-04-01    1945
2016-04-02    2007
2016-04-03    2313
2016-04-04    2147
2016-04-05     686
2016-04-06     168
2016-04-07      88
Name: date_crawled, dtype: int64

In [23]:
(autos["date_crawled"]
        .str[:10]
        .value_counts( dropna=False)
        .sort_values()
        )

2016-04-07      88
2016-04-06     168
2016-03-18     661
2016-04-05     686
2016-03-06     789
2016-03-13     922
2016-03-05    1366
2016-03-16    1506
2016-03-17    1559
2016-03-22    1596
2016-03-10    1609
2016-03-08    1658
2016-03-27    1658
2016-03-24    1662
2016-03-23    1692
2016-03-31    1725
2016-03-15    1729
2016-03-11    1735
2016-03-09    1740
2016-03-26    1750
2016-03-30    1768
2016-03-25    1780
2016-03-29    1816
2016-03-21    1819
2016-03-19    1837
2016-03-28    1888
2016-03-07    1892
2016-03-14    1916
2016-03-12    1940
2016-04-01    1945
2016-03-20    1956
2016-04-02    2007
2016-04-04    2147
2016-04-03    2313
Name: date_crawled, dtype: int64

Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.

In [24]:
(autos["last_seen"]
        .str[:10]
        .value_counts( dropna=False)
        .sort_index()
        )

2016-03-05       31
2016-03-06       81
2016-03-07      149
2016-03-08      180
2016-03-09      281
2016-03-10      374
2016-03-11      459
2016-03-12      684
2016-03-13      278
2016-03-14      452
2016-03-15      551
2016-03-16      553
2016-03-17     1009
2016-03-18      260
2016-03-19      543
2016-03-20      693
2016-03-21      688
2016-03-22      788
2016-03-23      711
2016-03-24      688
2016-03-25      685
2016-03-26      559
2016-03-27      566
2016-03-28      754
2016-03-29      871
2016-03-30      958
2016-03-31      970
2016-04-01      948
2016-04-02      982
2016-04-03     1080
2016-04-04     1052
2016-04-05     8876
2016-04-06    15928
2016-04-07     9641
Name: last_seen, dtype: int64

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

In [25]:
print(autos["ad_created"].str[:10].unique().shape)
(autos["ad_created"]
        .str[:10]
        .value_counts( dropna=False)
        .sort_index()
        )

(81,)


2015-08-07       1
2015-10-14       1
2015-11-08       1
2015-11-13       1
2015-12-28       1
              ... 
2016-04-03    2334
2016-04-04    2139
2016-04-05     628
2016-04-06     167
2016-04-07      84
Name: ad_created, Length: 81, dtype: int64

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.

In [26]:
autos["registration_year"].describe()

count    53323.000000
mean      2008.779045
std          6.838614
min       1929.000000
25%       2008.000000
50%       2010.000000
75%       2012.000000
max       2017.000000
Name: registration_year, dtype: float64

The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values. The minimum value is 1929, 2017


# Dealing with Incorrect Registration Year Data


Because a car can't be first registered before the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

One option is to remove the listings with these values. Let's determine what percentage of our data has invalid values in this column:

In [27]:
(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]

1.875363351649382e-05

Given that this is less than 1% of our data, we will remove these rows.

In [28]:
autos = autos[autos["registration_year"].between(1900,2016)]



In [29]:
autos["registration_year"].describe()

count    53322.000000
mean      2008.778891
std          6.838585
min       1929.000000
25%       2008.000000
50%       2010.000000
75%       2012.000000
max       2016.000000
Name: registration_year, dtype: float64

In [30]:
autos["registration_year"].value_counts().head(10)

2011    7236
2010    6064
2012    6056
2009    5289
2008    4949
2013    4506
2014    3813
2007    3671
2006    2769
2015    2342
Name: registration_year, dtype: int64

# Exploring Price by Brand

In [31]:
autos["brand"].value_counts(normalize=True)

volkswagen       0.197423
bmw              0.188346
audi             0.174487
mercedes_benz    0.158790
ford             0.035708
porsche          0.032069
opel             0.031544
mini             0.022805
skoda            0.019317
nissan           0.013297
seat             0.011890
hyundai          0.009640
volvo            0.009227
renault          0.009208
toyota           0.008514
mazda            0.008327
peugeot          0.008271
land_rover       0.007670
kia              0.007014
chevrolet        0.005457
fiat             0.005457
citroen          0.005139
jeep             0.004538
honda            0.003770
jaguar           0.003732
alfa_romeo       0.003432
mitsubishi       0.003357
suzuki           0.003019
chrysler         0.002269
dacia            0.001988
smart            0.001425
subaru           0.001407
saab             0.000713
lancia           0.000450
rover            0.000150
trabant          0.000075
lada             0.000038
daihatsu         0.000038
Name: brand,


German manufacturers represent four out of the top five brands, almost 50% of the overall listings. Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands combined.

There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 3% of total listing

In [32]:
brand_counts = autos["brand"].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .03].index
print(common_brands)

Index(['volkswagen', 'bmw', 'audi', 'mercedes_benz', 'ford', 'porsche',
       'opel'],
      dtype='object')


In [33]:
brand_mean_prices = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

{'volkswagen': 17376,
 'bmw': 19145,
 'audi': 19870,
 'mercedes_benz': 20664,
 'ford': 17121,
 'porsche': 51136,
 'opel': 14659}

Of the top 7 brands, there is a little variation in the price gap:

Audi, BMW, Ford, Volkswagen and Mercedes Benz are in the same price range

Porsche is a more expensive brand

Opel is the cheapest brand listed 

 # Exploring Mileage

In [34]:
bmp_series = pd.Series(brand_mean_prices)
pd.DataFrame(bmp_series, columns=["mean_price"])


mean_price
volkswagen          17376
bmw                 19145
audi                19870
mercedes_benz       20664
ford                17121
porsche             51136
opel                14659

In [35]:
brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

In [36]:
brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

mean_mileage
bmw                  108584
mercedes_benz        105020
audi                  99509
porsche               96555
volkswagen            84430
ford                  72938
opel                  63570

In [37]:
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  108584       19145
mercedes_benz        105020       20664
audi                  99509       19870
porsche               96555       51136
volkswagen            84430       17376
ford                  72938       17121
opel                  63570       14659

The range of car mileages varies much more than the prices do. There is a slight trend to the more expensive vehicles having higher mileage (except the porsche), with the less expensive vehicles having lower mileage.